From: Yu Watanabe <watanabe.yu+github@gmail.com>
Date: Sat, 21 Jun 2025 03:51:30 +0900
Subject: network/dhcp6: consider the DHCPv6 protocol as finished when conflict addresses exist

Replaces #37891.

Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/systemd/+bug/2115418
Origin: backport, https://github.com/systemd/systemd/commit/4c4fe8cd71da64506705736fda0ab7f7154f28d2
[chengen - Fix context in networkd-address.h / adjust content in networkd-dhcp6.c due to the absence of commit #03f065846096.]

--- systemd-255.4.orig/src/network/networkd-address.c
+++ systemd-255.4/src/network/networkd-address.c
@@ -267,6 +267,27 @@ bool link_check_addresses_ready(Link *li
                 has = true;
         }
 
+        if (has || source != NETWORK_CONFIG_SOURCE_DHCP6)
+                return has;
+
+        /* If there is no DHCPv6 addresses, but all conflicting addresses are successfully configured, then
+         * let's handle the DHCPv6 client successfully configured addresses. Otherwise, if the conflicted
+         * addresses are static or foreign, and there is no other dynamic addressing protocol enabled, then
+         * the link will never enter the configured state. See also link_check_ready(). */
+        SET_FOREACH(a, link->addresses) {
+                if (source >= 0 && a->source == NETWORK_CONFIG_SOURCE_DHCP6)
+                        continue;
+                if (!a->also_requested_by_dhcp6)
+                        continue;
+                if (address_is_marked(a))
+                        continue;
+                if (!address_exists(a))
+                        continue;
+                if (!address_is_ready(a))
+                        return false;
+                has = true;
+        }
+
         return has;
 }
 
@@ -1826,6 +1847,7 @@ int manager_rtnl_process_address(sd_netl
                 nft_set_context_clear(&address->nft_set_context);
                 (void) nft_set_context_dup(&a->nft_set_context, &address->nft_set_context);
                 address->requested_as_null = a->requested_as_null;
+                address->also_requested_by_dhcp6 = a->also_requested_by_dhcp6;
                 address->callback = a->callback;
         }
 
--- systemd-255.4.orig/src/network/networkd-address.h
+++ systemd-255.4/src/network/networkd-address.h
@@ -55,6 +55,10 @@ struct Address {
         bool scope_set:1;
         bool ip_masquerade_done:1;
         bool requested_as_null:1;
+        /* If this address is configured by non-DHCPv6 protocol (i.e. statically, NDISC, or any foreign
+         * methods), and the DHCPv6 client requests the same address, set this flag.
+         * See link_check_addresses_ready() and verify_dhcp6_address(). */
+        bool also_requested_by_dhcp6:1;
 
         /* duplicate_address_detection is only used by static or IPv4 dynamic addresses.
          * To control DAD for IPv6 dynamic addresses, set IFA_F_NODAD to flags. */
--- systemd-255.4.orig/src/network/networkd-dhcp6.c
+++ systemd-255.4/src/network/networkd-dhcp6.c
@@ -171,15 +171,21 @@ static int verify_dhcp6_address(Link *li
         if (existing->source == NETWORK_CONFIG_SOURCE_NDISC)
                 by_ndisc = true;
 
-        log_link_warning(link, "Ignoring DHCPv6 address %s/%u (valid %s, preferred %s) which conflicts with %s/%u%s.",
-                         pretty, address->prefixlen,
-                         FORMAT_LIFETIME(address->lifetime_valid_usec),
-                         FORMAT_LIFETIME(address->lifetime_preferred_usec),
-                         pretty, existing->prefixlen,
-                         by_ndisc ? " assigned by NDisc" : "");
+        log_link_full(link,
+                      /* Downgrade the log level if we have logged about the conflict previously. */
+                      existing->also_requested_by_dhcp6 ? LOG_DEBUG : LOG_WARNING,
+                      "Ignoring DHCPv6 address %s/%u (valid %s, preferred %s) which conflicts with %s/%u%s.",
+                      pretty, address->prefixlen,
+                      FORMAT_LIFETIME(address->lifetime_valid_usec),
+                      FORMAT_LIFETIME(address->lifetime_preferred_usec),
+                      pretty, existing->prefixlen,
+                      by_ndisc ? " assigned by NDisc" : "");
         if (by_ndisc)
-                log_link_warning(link, "Hint: use IPv6Token= setting to change the address generated by NDisc or set UseAutonomousPrefix=no.");
+                log_link_full(link,
+                              existing->also_requested_by_dhcp6 ? LOG_DEBUG : LOG_WARNING,
+                              "Hint: use IPv6Token= setting to change the address generated by NDisc or set UseAutonomousPrefix=no.");
 
+        existing->also_requested_by_dhcp6 = true;
         return -EEXIST;
 
 simple_log:
